### DISCUSSION TITLE
Display geo map on ImPlot canvas

### DISCUSSION THREAD

**Discussion by gorbatschow** [epezent/implot#431] [Show and tell]
I made a simple mapping application, it uses OpenStreetMap as tile provider.

The complete demo https://github.com/gorbatschow/ImOsmWidgetDemo
The source code of widget which renders tiles on ImPlot canvas  https://github.com/gorbatschow/ImOsmWidget

![image](https://raw.githubusercontent.com/gorbatschow/ImOsmWidgetDemo/master/doc/demo.gif)

**Comment by BenBE** [comment]
This looks great. :) Could you try to reduce the flickering while the new tiles are loaded? E.g. pre-loading the next detail level before scrolling or keeping the previews tiles for stretching until they can be replaced, may be two viable options (used on the slippy map on the web usually).

Also: Is the map provider abstracted away enough in a way such that you could use it for any tile source producing square images following the OSM tile naming scheme?

**Reply by epezent** [reply]
Very nice! We have a similar demo in the implot demos repo:

https://github.com/epezent/implot_demos/blob/master/demos/maps.cpp

**Reply by gorbatschow** [reply]
Thanks for the feedback. Yes, I'm thinking about making the map widget more sliding.
You can use any tile server even local. For offline applications normal practice is to use SQLite database as a tile storage. Such databases has special structure and called "mbtiles". I plan to add DB support too.

Oh, thanks, now I have seen your demo :)
The key difference is that in mine tiles are not stored on the disk and used memory always fits to the visible region.

**Reply by BenBE** [reply]
Re tile providers I more thought along the lines of some callback which is called for each required tile providing the x, y and zoom level needed and a buffer to write the tile into. Think along the lines of https://xkcd.com/195/ or other maps where tiles might just be generated on-the-fly in memory. Providing a default "Give me a tile server URL" handler and another "here's your DB" handler would be nice examples how that interface would work.

**Comment by aiekick** [comment]
Nice !
btw like the attention is on the center, could be better to download the new tiles in spiral from the center ?
especially when you dont have a fast internet
